torch/csrc/distributed/c10d/Hooks.hpp

```diff
@@ -22,8 +18,32 @@ struct TORCH_API EventInfo {
   int64_t drop_count;
 };
 
-// TODO do we want to expose something else here?
-TORCH_API bool dequeue_c10d_event(EventInfo& evt);
+typedef std::function<void(const EventInfo&)> CollectiveEventCallback;
+
+/**
+ * Register a callback that is invoked whenever a collective event happens.
+ *
+ * Locking:
+ *  Registration takes a subsystem specific lock.
+ *  callback invocation happens when the same lock held.
+ *
+ * Callbacks must not block or run for a long period of time.
+ * They are invoked from threads are part of PyTorch's critical distributed
+ * infrastrucute. The recomended pattern is for callbacks to enqueue the events
+ * on some queue and have a separate thread process those events. If the
+ * callback deadlocks, it will hang the whole process and stop PyTorch from
+ * detecting failures. Do not call into CUDA.
```

Comment:
Well, so true what if python thread hangs at GIL...

Reply:
this comment refers to the c++ callback. the c++ callback must not hang, or it hangs our watchdog and stops our timeout mechanism.  That's why its pretty much only allowed to put events onto a queue.

the thread that reads from the queue (whether thats a c++ thread or a python thread) can hang if it wants to- that wouldn't affect our watchdog, it just affects whatever system is built on top of that callback.